src/server/job_func.c

```diff
@@ -685,7 +685,7 @@ remove_stdouterr_files(job *pjob, char *suffix) {
  */
 int direct_write_requested(job *pjob) {
 	char *pj_attrk = NULL;
-	if ((pjob->ji_wattr[(int)JOB_ATR_keep].at_flags & ATR_VFLAG_SET)) {
+	if (pjob && (pjob->ji_wattr[(int)JOB_ATR_keep].at_flags & ATR_VFLAG_SET)) {
```

Comment:
While this is a fine check to add, shouldn't you really be returning after the find_job() in req_copyfile().  Is there any purpose to continue processing that function if we can't find the job?

Reply:
@bhroam This is really a corner case. I think instead of doing something special in this function we should add this check to the caller and take the decision of bailing out from there.
What do you think @nithinj?

Reply:
I think we can continue processing the stage out even if find_job didn't returned anything as rq_cpyfile structure contains stage out info.

Reply:
I'm not sure of the ramifications of the situation we'd be in.  What does it mean to the mom if she can't find the job?  While we might be able to continue with the rq_cpyfile structure, would anything in there be valid?  Would the files still exist any more?  Would Mom have cleaned up?

Reply:
I looked at the callers of direct_write_requested(). One of them is req_cpyfile(). That functions does sections of code that is different when pjob is NULL vs not-NULL. Thus, it is clear, that some part of the copyfile processing is expected to continue even when pjob is NULL.

In the WIN32 counterpart req_cpyfile(), this function is called after a check for pjob being NON-NULL, thus:
      if (pjob != NULL && (dir == STAGE_DIR_OUT)) {
		direct_write = direct_write_requested(pjob);
	}

We could simply mimic this behavior for Linux. 

But then this function is also called from "open_std_out_err", and if this function fails, the code will go ahead to print some properties of the job as part of the error message (again access pjob).  However, this function "open_std_out_err" is called from functions where pjob CANNOT be NULL, so in effect, the caller is confident :)

The bigger issues is this: direct_write_requested() returns 0 if direct write is not requested by job, and 1 is direct write was requested. If pjob passed to it is NULL, and you handle it inside this function, what would you return? Either 0, or 1 would be a wrong return. If you add a different return value for pjob being NULL, then you anyway have to add code to the caller. 

Given all these above, I strongly suggest we put a check in the caller and do the "right" thing for that caller.

Reply:
We rely on pjob structure to determine whether direct_write is requested for the job, not rq_cpyfile. If the answer is no, or if it is something that cannot be determined the function will return false. In the worst case where the job has requested direct write and direct_write_requested function returning false will result in warning in mom log when it tries to stageout/delete stdout/err files from spool directory. This will still be the case if we handle the NULL case in the caller. But the issue being discussed is whether we should return from req_cpyfile() itself if the mom cannot find the job.

`What does it mean to the mom if she can't find the job? While we might be able to continue with the rq_cpyfile structure, would anything in there be valid? Would the files still exist any more? Would Mom have cleaned up?`
I still doesn't know the answer to these questions. I might need more time to investigate this (missing the bigger picture). I think @bayucan will take over this issue as I'm working on another one. Thanks for all the ideas!